First version of TPC QC Tracks task

stheckel opened the pull request:
o This is the O2 part of the TPC QC task relating to the tracks
o Several 1D and 2D histograms are implemented
o A root macro is provided to run the task and save the output
o A corresponding workflow will be added to QualityControl

stheckel commented:
Dear @wiechula,

please have a look at the first version of the O2 TPC QC tracks task. Some additional info:
- As discussed, I have also added a macro to run the task and save the output.
- I plan to add the nice style (drawing option, no stats on 2D histograms) from the helper functions provided by Thomas as soons as his pr is merged.
- At the moment, one hard coded cut is used, which is a minimum number of TPC clusters of 20. In future, I also plan to use Thomas' cut class. For the pre-commissioning, we also could consider not to take any cuts at all.

Cheers,
Stefan

wiechula commented on Detectors/TPC/qc/include/TPCQC/Tracks.h:
```diff
@@ -0,0 +1,82 @@
+{
+ public:
+  // default constructor
+  Tracks();
```
Please use
```c++
Tracks() = default;
```
and remove the implementation in the cxx

wiechula commented on Detectors/TPC/qc/include/TPCQC/Tracks.h:
```diff
@@ -0,0 +1,82 @@
+
+ private:
+  std::vector<TH1F> mHist1D;
+  std::vector<TH2F> mHist2D;
```
Default initialize both vectors adding `{}` at the end.
Add a doxygen comment with `///< comment`

wiechula commented on Detectors/TPC/qc/macro/runTracks.C:
```diff
@@ -0,0 +1,122 @@
+
+  // ===| get histograms |=======================================================
+  auto histos1D = tracksQC.getHistograms1D();
+  auto histos2D = tracksQC.getHistograms2D();
```
I suggest to use
```c++
const auto& ...
```
in both cases. Otherwise you'll copy all histos.

wiechula commented on Detectors/TPC/qc/src/Tracks.cxx:
```diff
@@ -0,0 +1,126 @@
+
+Tracks::Tracks() : mHist1D{}, mHist2D{}
+{
+}
```
Remove implementation of ctor, see comments in header.

stheckel commented:
Hi @wiechula,

I am working on your suggested changes and in addition I want to use the helpers functions provided by Thomas, which are now merged to dev. In principle, this works, but in the `runTracks.C` I cannot make the histograms `const auto&`, because then the draw function does not work anymore. It expects a non-const object. What indeed works, is `auto&` or `auto&&`, where both correspond to a non-const reference, but I do not fully get the difference. What I found is, that the latter allows for a temporary initializer, which is not the case here, if I understand correctly. So both would be fine. Please tell me your preference or if you have any other idea.

Cheers,
Stefan

wiechula commented:
> Hi @wiechula,
> 
> I am working on your suggested changes and in addition I want to use the helpers functions provided by Thomas, which are now merged to dev. In principle, this works, but in the `runTracks.C` I cannot make the histograms `const auto&`, because then the draw function does not work anymore. It expects a non-const object. What indeed works, is `auto&` or `auto&&`, where both correspond to a non-const reference, but I do not fully get the difference. What I found is, that the latter allows for a temporary initializer, which is not the case here, if I understand correctly. So both would be fine. Please tell me your preference or if you have any other idea.
> 

Sure, you can't modify a `const ` object. So use `auto&` without const.

> Interstingly, in both cases (`auto&` and `auto&&`), the draw options are only applied to the histogram drawn on the pads of the canvas, the bare histograms saved later in the `dumpToFile` function do not contain any of these changes. This is fine, but I am wondering a bit, why `draw` needs a non-const object, if it anyhow does not change the object itself, but draws it's own version of it...

Hmm, this I also don't understand. Ok, the `Draw` function in the histogram class is not `const` so it cannot be used on `const` objects. Funny this never occurred to me before...
But the changes should be stored in the file I think. Can you upload a new version, then I can have a look.

stheckel commented:
Hi Jens,

I just pushed the changes requested by you and I now also use the helpers to beautify the histograms. Actually, having a closer look I now see, that the 2D histograms indeed **are** changed by calling draw. Just the logy in the 1D histogram somehow is not stored.

Cheers,
Stefan

wiechula commented on Detectors/TPC/qc/CMakeLists.txt.orig:
```diff
@@ -0,0 +1,79 @@
+#Copyright CERN and copyright holders of ALICE O2.This software is distributed
```
This file must be deleted! Please be careful in what is committed and pushed.

wiechula commented on Detectors/TPC/qc/include/TPCQC/Tracks.h:
```diff
@@ -0,0 +1,83 @@
+ private:
+  /// Initialize vectors of 1D and 2D histograms
+  std::vector<TH1F> mHist1D{};
+  std::vector<TH2F> mHist2D{};
```
for comments of data memebers please use
`///< comment` after the data member.

wiechula commented on Detectors/TPC/qc/src/TPCQCLinkDef.h.orig:
```diff
@@ -0,0 +1,28 @@
+// Copyright CERN and copyright holders of ALICE O2. This software is
```
Also this file must be deleted!

wiechula commented on Detectors/TPC/qc/src/Tracks.cxx:
```diff
@@ -0,0 +1,122 @@
+}
+
+//______________________________________________________________________________
+void Tracks::dumpToFile(const std::string filename)
```
Are you sure the code compiles and links? In the header you use:
```c++
void dumpToFile(std::string filename);
```
here you use
```c++
void Tracks::dumpToFile(const std::string filename)
```
so `const std::string` there are different things. Anyway, I propose to use
```c++
void dumpToFile(std::string_view filename);
```
and then
```c++
auto f = std::unique_ptr<TFile>(TFile::Open(filename.data(), "recreate"));
```

stheckel replied:
Yes, this compiles. Actually I copied both lines from Thomas PID class, which is in exactly this way already part of dev. ;)
Still, you are right, this should be consistent. But I think, that his is exactly the place, where the string_view did not work for Thomas, so I would stay with a std::string. Should I make it const?

wiechula replied:
please try with `std::string_view` as described above. Thomas' problem with in the Clusters task.

wiechula replied:
And calling the `dumpToFile` function also works? I guess you checked that histograms are properly written.

wiechula commented on Detectors/TPC/qc/include/TPCQC/Tracks.h:
```diff
@@ -0,0 +1,83 @@
+  void resetHistograms();
+
+  /// Dump results to a file
+  void dumpToFile(std::string filename);
```
see comment in cxx file

stheckel commented:
Hi Jens,

I have deleted the two files, which should never have made it here. Sorry again for that, I will be more careful in the future. Furthermore, I have implemented the other little changes, including the std::string_view. You are right, it actually works here. I will tell Thomas to change this also in the PID task. Still, even if not really clean code, it was also working before, I checked that calling the `dumpToFile` worked and the histograms were stored.

Cheers,
Stefan

wiechula commented on Detectors/TPC/qc/include/TPCQC/Tracks.h:
```diff
@@ -0,0 +1,82 @@
+#ifndef AliceO2_TPC_QC_TRACKS_H
+#define AliceO2_TPC_QC_TRACKS_H
+
+#include <vector>
```
```c++
#include <string_view>
```

wiechula commented on Detectors/TPC/qc/src/Tracks.cxx:
```diff
@@ -0,0 +1,122 @@
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+#include <cmath>
```
I don't see anything used from `cmath` in this class. So you could remove it.

wiechula commented on Detectors/TPC/qc/src/Tracks.cxx:
```diff
@@ -0,0 +1,122 @@
+//______________________________________________________________________________
+void Tracks::dumpToFile(std::string_view filename)
+{
+  auto f = std::unique_ptr<TFile>(TFile::Open(filename.data(), "recreate"));
```
I think using `unique_ptr` only works because root includes it somewhere. To be more explicit I suggest to add
```c++
#include <memory>
```
above

stheckel commented:
No problem, I have made the changes. Now I also have one: I checked, and Thomas is using `M_PI` for the ranges in phi instead of simply 3.14. As I think, that this is better, I have changed it accordingly. Now the question is: shall I then keep the cmath include? It also works without it, and in principle the `M_PI` is provided by the `#define _USE_MATH_DEFINES`, but I am not sure, whether one still should also keep the `#include <cmath>`.

wiechula commented:
Yes, then put both
```c++
#define _USE_MATH_DEFINES
#include <cmath>
```

stheckel commented:
Ok, both lines are in the current version here.

One more question: I have now two commits, the original one with some amendments and the second, using the helpers and also some minor modifications. Now I see, that the two `*.orig` files from the merging have been introduced already with an amendment to the first commit, but they are deleted only in the second. Would it then be better to merge also these two commits (i.e. via `git rebase -i`)? In the end, the question is: If I add and delete a file in a single commit, does it then appear at all in the commit history or can we get rid of it completely in this way?

And one suggestion: Maybe one could add to the automatic code checking, if such files are included a commit.

Cheers,
Stefan

stheckel commented:
Hi @wiechula,

the checks have been successful. Do you have any further remarks or shall I make the pull request ready for review?

Cheers,
Stefan

wiechula commented:
Yes, please remove [WIP] and chage to normal PR

shahor02 commented on Detectors/TPC/qc/src/Tracks.cxx:
```diff
@@ -0,0 +1,125 @@
+  if (eta > 0.) {
+    mHist1D[3].Fill(phi);
+  }
+  if (eta < 0.) {
```
why not using ``else``?

shahor02 commented on Detectors/TPC/qc/src/Tracks.cxx:
```diff
@@ -0,0 +1,125 @@
+  if (eta > 0.) {
+    mHist2D[1].Fill(phi, nCls);
+  }
+  if (eta < 0.) {
```
idem.

shahor02 commented on Detectors/TPC/qc/src/Tracks.cxx:
```diff
@@ -0,0 +1,125 @@
+  if (sign < 0.) {
+    mHist2D[5].Fill(phi, eta);
+  }
+  if (sign > 0.) {
```
idem

wiechula commented:
@shahor02 , I totally agree. Thanks!

stheckel commented:
Dear @shahor02, dear @wiechula,

yes, good point, I will change this. One more question, maybe more to Jens: shall I make it `else` or `else if`? In the current implementation we explicitly do not take tracks at `eta=0` or with `sign=0`. While in case of `eta`, this will probably be no issue, in case of `sign` it could make a difference. Later, in physics running, we should always have the sign, but during pre-comissioning there is no field. If we use `else`, we would fill also `sign=0` tracks in the histograms for positive charge. I would like to clarify, if we consider this a feature or rather would like to avoid it.

Cheers,
Stefan

shahor02 commented:
@stheckel for the sign: there should be no sign=0 even in B=0 runs: we will always assign a typical momentum to account for the mult. scattering. So, ``if (sign>0) {...} else {...}`` should be ok.
But this is indeed a good point (just discussed with @davidrohr and pinging to @iouribelikov and @mpuccio ): we should make sure that with B=0 field the "typical" ``q/pT`` is indeed assigned.

wiechula commented:
@stheckel , I agree with Ruben. Just use if ... else

iouribelikov commented:
@shahor02 Hi Ruben, everybody.
If I remember correctly, the argument for assigning to tracks "some typical" momentum was the necessity to calculate "some effective" multiple scattering (even in case when B=0).

shahor02 commented:
@iouribelikov yes, this is what I wrote above.

iouribelikov commented:
@shahor02  Sorry, I am sitting at the ITS3 meeting... Indeed you already mentioned it,  and no additional comment was really needed.  Cheers. Yura.

stheckel commented:
@wiechula, @shahor02,
I have made the changes, pushed again and the checks are all successful. Any further comments or shall I make the pr ready for review?
Cheers, Stefan

shahor02 commented:
Hi @stheckel 
for me it is ok.
Cheers,
 Ruben

stheckel commented:
Hi @shahor02,
thanks! As Jens was fine with it before and also agreed to the last change, I think the pr could be merged now.
Cheers,
Stefan

davidrohr commented:
> @stheckel for the sign: there should be no sign=0 even in B=0 runs: we will always assign a typical momentum to account for the mult. scattering. So, `if (sign>0) {...} else {...}` should be ok.
> But this is indeed a good point (just discussed with @davidrohr and pinging to @iouribelikov and @mpuccio ): we should make sure that with B=0 field the "typical" `q/pT` is indeed assigned.

Fix is here: #2746
